File: examples/python/agent/hello.py
import logging
import os

from rebuno import Agent, execution, tool

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(name)s %(levelname)s %(message)s"
)
logger = logging.getLogger("hello-agent")

agent = Agent(
    "hello",
    kernel_url=os.environ.get("REBUNO_KERNEL_URL", "http://localhost:8001"),
)


@tool("reverse")
def reverse(text: str) -> dict:
    """Reverse a string."""
    return {"reversed": text[::-1]}


@tool("word_count")
def word_count(text: str) -> dict:
    """Count words in a string."""
    return {"count": len(text.split())}


async def process(query: str) -> dict:
    logger.info("Processing query: %s (execution=%s)", query, execution.id)
    reversed_result = await reverse(text=query)
    count_result = await word_count(text=query)

    logger.info(
        "Agent finished: reversed=%s count=%s",
        reversed_result["reversed"], count_result["count"],
    )
    return {
        "query": query,
        "reversed": reversed_result["reversed"],
        "word_count": count_result["count"],
    }


if __name__ == "__main__":
    logger.info("Starting hello agent")
    agent.run(process)

File: examples/typescript/agent/hello.ts
import { z } from "zod";
import { BaseAgent, AgentContext, defineTool } from "rebuno";

const agent = new (class HelloAgent extends BaseAgent {
  async process(ctx: AgentContext) {
    const input = ctx.input as Record<string, unknown> | string;
    const query =
      typeof input === "object" && input !== null
        ? (input.query as string) ?? ""
        : String(input);

    console.log(`Processing query: ${query}`);

    const reverseResult = (await ctx.invokeTool("reverse", {
      text: query,
    })) as Record<string, unknown>;
    const countResult = (await ctx.invokeTool("word_count", {
      text: query,
    })) as Record<string, unknown>;

    console.log(
      `Agent finished: reversed=${reverseResult.reversed}, count=${countResult.count}`,
    );

    return {
      query,
      reversed: reverseResult.reversed,
      wordCount: countResult.count,
    };
  }
})({
  agentId: "hello",
  kernelUrl: process.env.REBUNO_KERNEL_URL ?? "http://localhost:8080",
});

agent.addTool(
  defineTool({
    id: "reverse",
    description: "Reverse a string.",
    inputSchema: z.object({ text: z.string() }),
    execute: async ({ text }) => ({ reversed: text.split("").reverse().join("") }),
  }),
);

agent.addTool(
  defineTool({
    id: "word_count",
    description: "Count words in a string.",
    inputSchema: z.object({ text: z.string() }),
    execute: async ({ text }) => ({ count: text.split(/\s+/).filter(Boolean).length }),
  }),
);

console.log("Starting hello agent");
agent.run();
